src/json/def_to_json.cr

```diff
@@ -0,0 +1,186 @@
+module JSON
+  # The `def_to_json` macro generates a `to_json(json : JSON::Builder)` method based on provided *mappings*.
+  #
+  # It is a lightweight alternative to `JSON.mapping` if you don't need to declare instance variables and a parser.
+  #
+  # The generated method invoks `to_json(JSON::Builder)` on each of the values returned by the *value* expression,
+  # or - if a converter is specified - `to_json(value, JSON::Builder)` on the converter.
+  #
+  # ### Example
+  #
+  # ```
+  # require "json"
+  #
+  # record Location, lat : Float64, long : Float64 do
+  #   JSON.def_to_json([lat, long])
+  # end
+  #
+  # record House, street : String, street_number : Int32, location : Location do
+  #   JSON.def_to_json(
+  #     address: _,
+  #     loc: location,
+  #     empty_field: {emit_null: true},
+  #     next_number: {value: street_number + 1},
+  #   )
+  #
+  #   def address
+  #     "#{street} #{street_number}"
+  #   end
+  #
+  #   def empty_field
+  #     nil
+  #   end
+  # end
+  #
+  # house = House.new("Crystal Road", 1234, Location.new(12.3, 34.5))
+  # house.to_json # => %({"address":"Crystal Road 1234","loc":{"lat":12.3,"long":34.5},"empty_field":null,"next_number":1235})
+  # ```
+  #
+  # ### Usage
+  #
+  # `JSON.def_to_json` must receive a series of named arguments, or a named tuple literal, or a hash literal,
+  # whose keys will define JSON properties.
+  #
+  # The value of each key can be a hash or named tuple literal with the following options:
+  # * **value**: the Crystal expression to determine the value. By default it is equal to the property name of the current  *key*
+  #   on the Crystal object (as opposed to the key in the JSON document)
+  # * **emit_null**: if `true`, emits a `null` value if the *value* is `nil` (by default nulls are not emitted)
+  # * **converter**: specify an alternate type for generation. The converter must define `to_json(value, JSON::Builder)` as class methods. Examples of converters are `Time::Format` and `Time::EpochConverter` for `Time`.
+  # * **root**: assume the value is inside a JSON object with a given key
+  #
+  # If it is not a hash or named tuple literal, the expression will be interpreted as `value` parameter. As a shortcut `_` represents
```

Comment:
I'm not sure about the `_` shortcut, I think `foo: foo` is perhaps the clearest. I can see it getting terrible with long names though.